pv3_rust/src/pv.rs

```diff
@@ -0,0 +1,415 @@
+            received = metadata_count;
+        }
+    }
+
+    unsafe {
+        if xsk_ring_prod__needs_wakeup(&mut nic.fq) != 0 {
+            libc::recvfrom(xsk_socket__fd(nic.xsk), std::ptr::null_mut::<c_void>(), 0, libc::MSG_DONTWAIT, std::ptr::null_mut::<libc::sockaddr>(), std::ptr::null_mut::<u32>()) ;
+        }
+    }
+
+    received
+}
+
+pub fn pv_send(nic: &mut PvNic, packets: &mut Vec<PvPacket>, batch_size: u32) -> u32 {
+    /* free packet metadata and UMEM chunks as much as the number of filled slots in cq. */
+    let mut cq_idx: u32 = 0;
+    let filled: u32 = unsafe{ xsk_ring_cons__peek(&mut nic.cq, batch_size, &mut cq_idx) }; // fetch the number of filled slots(the number of packets completely sent) in cq
+
+    if filled > 0 {
+        for _ in 0..filled {
+            unsafe { pv_free_(nic, *xsk_ring_cons__comp_addr(&nic.cq, cq_idx)); } // free UMEM chunks as much as the number of sent packets (same as **filled)
+            cq_idx += 1;
+        }
+        unsafe { xsk_ring_cons__release(&mut nic.cq, filled); } // notify kernel that cq has empty slots with **filled (Dequeue)
+    }
+
+     /* reserve TX ring as much as batch_size before sending packets. */
+    let mut tx_idx: u32 = 0;
+    let reserved: u32 = unsafe { xsk_ring_prod__reserve(&mut nic.tx, batch_size, &mut tx_idx) };
+    // println!("packets len: {}, tx reserved: {}, tx_idx: {}", packets.len(), reserved, tx_idx);
+
+    /* send packets if TX ring has been reserved with **batch_size. if not, don't send packets and free them */
+    if reserved < batch_size {
+        /* in case that kernel lost interrupt signal in the previous sending packets procedure,
+        repeat to interrupt kernel to send packets which ketnel could have still held.
+        (this procedure is for clearing filled slots in cq, so that cq can be reserved as much as **batch_size in the next execution of pv_send(). */
+        unsafe{
+            if xsk_ring_prod__needs_wakeup(&mut nic.tx) != 0 {
+                libc::sendto(xsk_socket__fd(nic.xsk), std::ptr::null(), 0, libc::MSG_DONTWAIT, std::ptr::null(), 0);
+            }
+        }
+        return 0;
+    }
+
+    /* send packets */
+    let mut index_count: u32 = 0;
+    for i in (0..reserved).rev() {
+        /* insert packets to be send into TX ring (Enqueue) */
+        let rx_desc_ptr = unsafe { xsk_ring_prod__tx_desc(&mut nic.tx, tx_idx + index_count) } ;
+        unsafe {
+            rx_desc_ptr.as_mut().unwrap().addr = packets[i as usize].private as u64 + packets[i as usize].start as u64;
+            rx_desc_ptr.as_mut().unwrap().len = packets[i as usize].end - packets[i as usize].start;
+            // println!("ptr: {:?}, addr: {}, len: {}", rx_desc_ptr, rx_desc_ptr.as_mut().unwrap().addr, rx_desc_ptr.as_mut().unwrap().len);
+        }
+
+        // packet_dump(&packets[i as usize]);
+        packets.pop();   // free packet metadata of sent packets.
+        index_count += 1;
+    }
```

Comment:
```suggestion
    for i in 0..reserved.rev() {
        let pkt_index: usize = (reserved - 1) as usize;
        /* insert packets to be send into TX ring (Enqueue) */
        let rx_desc_ptr = unsafe { xsk_ring_prod__tx_desc(&mut nic.tx, tx_idx + i) } ;
        unsafe {
            rx_desc_ptr.as_mut().unwrap().addr = packets[pkt_index].private as u64 + packets[pkt_index].start as u64;
            rx_desc_ptr.as_mut().unwrap().len = packets[pkt_index].end - packets[pkt_index].start;
            // println!("ptr: {:?}, addr: {}, len: {}", rx_desc_ptr, rx_desc_ptr.as_mut().unwrap().addr, rx_desc_ptr.as_mut().unwrap().len);
        }

        // packet_dump(&packets[pkt_index]);
        packets.pop();   // free packet metadata of sent packets.
    }
```

Reply:
the suggested code is better than before.

Reply:
`pkt_index` can be lower than 0.

Reply:
My bad. `pkt_index` doesn't go down below 0